include/maths/CBoostedTreeLeafNodeStatistics.h

```diff
@@ -37,14 +45,300 @@ class CEncodedDataFrameRowRef;
+                }
+                m_MissingDerivatives[i].add(other.m_MissingDerivatives[i]);
+            }
+        }
+
+        //! Subtract \p rhs.
+        void subtract(const CPerSplitDerivatives& rhs) {
+            for (std::size_t i = 0; i < m_Derivatives.size(); ++i) {
+                for (std::size_t j = 0; j < m_Derivatives[i].size(); ++j) {
+                    m_Derivatives[i][j].subtract(rhs.m_Derivatives[i][j]);
+                }
+                m_MissingDerivatives[i].subtract(rhs.m_MissingDerivatives[i]);
+            }
+        }
+
+        //! Remap the accumulated curvature to lower triangle row major format.
+        void remapCurvature() {
+            for (std::size_t i = 0; i < m_Derivatives.size(); ++i) {
+                for (std::size_t j = 0; j < m_Derivatives[i].size(); ++j) {
+                    m_Derivatives[i][j].remapCurvature();
+                }
+                m_MissingDerivatives[i].remapCurvature();
+            }
+        }
+
+        //! Get the memory used by this object.
+        std::size_t memoryUsage() const {
+            return core::CMemory::dynamicSize(m_Derivatives) +
+                   core::CMemory::dynamicSize(m_MissingDerivatives) +
+                   core::CMemory::dynamicSize(m_Storage);
+        }
+
+        //! Estimate the split derivatives' memory usage for a data frame with
+        //! \p numberCols columns using \p numberSplitsPerFeature for a loss
+        //! function with \p numberLossParameters parameters.
+        static std::size_t estimateMemoryUsage(std::size_t numberFeatures,
+                                               std::size_t numberSplitsPerFeature,
+                                               std::size_t numberLossParameters) {
+            std::size_t derivativesSize{numberFeatures * (numberSplitsPerFeature + 1) *
+                                        sizeof(CDerivatives)};
+            std::size_t storageSize{numberFeatures * (numberSplitsPerFeature + 1) * numberLossParameters *
+                                    (numberLossParameters + 1) * sizeof(double)};
+            return sizeof(CPerSplitDerivatives) + derivativesSize + storageSize;
+        }
+
+        //! Get a checksum of this object.
+        std::uint64_t checksum(std::uint64_t seed = 0) const {
+            seed = CChecksum::calculate(seed, m_NumberLossParameters);
+            seed = CChecksum::calculate(seed, m_Derivatives);
+            seed = CChecksum::calculate(seed, m_MissingDerivatives);
+            return seed;
+        }
+
+    private:
+        using TDerivativesVecVec = std::vector<TDerivativesVec>;
+
+    private:
+        static std::size_t number(const TDerivativesVec& derivatives) {
+            return derivatives.size();
+        }
+        static std::size_t number(const TImmutableRadixSet& splits) {
+            return splits.size() + 1;
+        }
+        template<typename SPLITS>
+        void map(const SPLITS& splits) {
```

Comment:
Can you please add a comment for posterity what this function is doing? It seems to have non-trivial business logic.

Reply:
Good idea, see [e1556ce](https://github.com/elastic/ml-cpp/pull/993/commits/e1556ce641f9804d8dfbb78142d0d69124ba4035).